transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java

```diff
@@ -744,7 +744,15 @@ private void invokeRead() {
         try {
             ((ChannelOutboundHandler) handler()).read(this);
         } catch (Throwable t) {
-            notifyHandlerException(t);
+            invokeExceptionCaughtFromOutbound(t);
```

Comment:
Aside: Hmmm... Why is `read()` in `ChannelOutboundHandler`? That seems just as wrong as exceptionCaught being on `ChannelInboundHandler`.

Reply:
because `read()` is really an outbound operation.... It tells the `Channel` we want to read something once it is ready... Just like calling `flush()` means we want to write everything that was written before as soon as the `Channel` is writable.

Reply:
read() is outbound from the pipeline's perspective, but not the Handler. Just like exceptionCaught is inbound from the pipeline's perspective, but not the Handler. `read()` interacts with channelRead and channelReadComplete. A ChannelInboundHandler shouldn't need to become a ChannelOutboundHandler because it wants to interact with outbound flow control.

Ah, shoot. It seems `channelWritabilityChanged` is wrong too. A ChannelOutboundHandler observing flow control has to be a ChannelInboundHandler as well...